batch/test/test_batch.py

```diff
@@ -1729,3 +1729,19 @@ def test_region(client: BatchClient):
     assert status['state'] == 'Success', str((status, b.debug_info()))
     assert status['status']['region'] == region, str((status, b.debug_info()))
     assert region in j.log()['main'], str((status, b.debug_info()))
+
+
+@pytest.mark.timeout(10 * 60)
+def test_for_siwei_bug_13399(client: BatchClient):
+    try:
+        bb = create_batch(client)
+        bb.create_job(DOCKER_ROOT_IMAGE, ['sleep', '300'])
+        bb.submit()
+
+        j = bb.create_job(DOCKER_ROOT_IMAGE, ['true'])
```

Comment:
This should make sure the job has been scheduled and the JOB_TOKEN_CACHE has a value in it for this range of jobs.

Reply:
The issue is the sub-query in the cache. Consider the cache query:
```
SELECT
batch_bunches.start_job_id,
batch_bunches.token,
(SELECT start_job_id FROM batch_bunches WHERE batch_id = %s AND start_job_id > %s ORDER BY start_job_id LIMIT 1) AS next_start_job_id,
batches.n_jobs
FROM batch_bunches
JOIN batches ON batches.id = batch_bunches.batch_id
WHERE batch_bunches.batch_id = %s AND batch_bunches.start_job_id <= %s
ORDER BY batch_bunches.start_job_id DESC
LIMIT 1;
```
Because we use `LIMIT 1`, we don't gap lock records with larger `start_job_id` (see [SO post](https://stackoverflow.com/a/5758358/6823256) that verifies, but easy for us to confirm).

We need this to happen:
1. Bunch 1 fully commits itself.
2. Bunch 1 job 1 tries to schedule and starts by executing the sub query, which read locks the last bunch (which is Bunch 1).
3. Bunch 2 inserts into `batch_bunches` and calls `/commit` (super fast!). This increments `batches.n_jobs` to account for Bunch 2's jobs.
4. Bunch 1 job 1 now executes the outer sub-query, grabbing a fresh value for `n_jobs` which includes Bunch 2's jobs.
5. Bunch 1 job 1 inserts a cache entry that spans both Bunch 1 and Bunch 2.
6. Bunch 2 job 1 tries to schedule and retrieves from the cache Bunch 1's specs.

Reply:
So, to reproduce, we need to load the database enough that that sub-query happens before the insert into batch_bunches while the outer query happens after we call commit.

I think our best bet to replicate that is to submit tons of fast-update bunches. The fast-update ensures we don't do network requests (so the insert into batch_bunches and the following commit are issued rapidly). And I think we just need to get lucky that some subsequent bunch zips in. It probably helps if the bunches are small so that they can be inserted rapidly.

Reply:
Heh 100 bunches can be created and executed in 0.11s. No bug triggered. I'm going to radically kick up the number of bunches.